### PR TITLE
Changes to support PETSc 3.21 (to be released March 29th) and single precision builds

### PR DESCRIPTION
**Pull request by stefanozampini**
The only change needed is a command line option in the tests.

I have added  a missing 3D Hcurl test (fixed in PETSc main) see https://github.com/mfem/mfem/issues/4050

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4203](https://github.com/mfem/mfem/pull/4203) | @stefanozampini | @tzanio | @jandrej + @sebastiangrimberg | 3/19/24 | 4/8/24 | 4/8/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Review by jandrej — COMMENTED** [review]
LGTM

@tzanio as the title suggests, don't merge before next petsc release

**Comment by stefanozampini** [comment]
This can be merged even now, no need to wait for 3.21

**Comment by tzanio** [comment]
@stefanozampini, is it `ready-for-review`?

**Comment by stefanozampini** [comment]
yes

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by tzanio — APPROVED** [review]
Thanks @stefanozampini !

File: examples/petsc/makefile [hunk 1]
```diff
@@ -83,7 +83,8 @@ EX1_ARGS_HIPAMG       := -m ../../data/star.mesh --usepetsc --device hip --petsc
 EX2_ARGS              := -m ../../data/beam-quad.mesh --usepetsc --petscopts rc_ex2p
 EX2_ARGS_BDDC         := -m ../../data/beam-tri.mesh --usepetsc --nonoverlapping --petscopts rc_ex2p_bddc
 EX2_ARGS_ASM          := -m ../../data/beam-quad.mesh --usepetsc --petscopts rc_ex2p_asm
-EX3_ARGS              := -m ../../data/klein-bottle.mesh -o 2 -f 0.1 --usepetsc --petscopts rc_ex3p_bddc --nonoverlapping
+EX3_ARGS_BDDC_2D      := -m ../../data/klein-bottle.mesh -o 2 -f 0.1 --usepetsc --petscopts rc_ex3p_bddc --nonoverlapping
```
[the remark this answers is not in the document]
Comment (stefanozampini) [reply examples/petsc/makefile:86]: done

**Review by jandrej — APPROVED** [review]

File: linalg/petsc.cpp [hunk 2]
```diff
@@ -3951,10 +3951,13 @@ void PetscH2Solver::H2SolverConstructor(ParFiniteElementSpace *fes)
...
-   PCSetCoordinates(*this,sdim,c.Size()/sdim,(PetscReal*)mfem::Read(c.GetMemory(),
-                                                                    c.Size(),false));
-   PCSetFromOptions(*this);
+
+   PC pc = (PC)obj;
+   ierr = PCSetType(pc,PCH2OPUS); PCHKERRQ(obj, ierr);
+   ierr = PCSetCoordinates(pc,sdim,c.Size()/sdim,
+                           (PetscReal*)mfem::Read(c.GetMemory(),
+                                                  c.Size(),false));
+   ierr = PCSetFromOptions(pc); PCHKERRQ(obj, ierr);
```
[the remark this answers is not in the document]
Comment (jandrej) [reply linalg/petsc.cpp:3960]: ping @stefanozampini

[the remark this answers is not in the document]
Comment (stefanozampini) [reply linalg/petsc.cpp:3960]: It is already guarded.  We were only ignoring the error code

**Comment by tzanio** [comment]
Now that PR https://github.com/mfem/mfem/pull/3922 has been merged in `master`, please update this PR by:

1. merging `master` in here and resolving conflicts
2. replacing use of `double` with `real_t`
3. guarding double/single-only sections with `MFEM_USE_DOUBLE` and `MFEM_USE_SINGLE`
4. making sure tests run cleanly with default and `MFEM_PRECISION=single` configurations

See https://github.com/orgs/mfem/discussions/4207 for more details and/or ask @dylan-copeland for help.

**Comment by stefanozampini** [comment]
@tzanio I have added support for single precision. When building using a `user.mk` file (the method used when PETSc builds MFEM) I have noticed that I have to specify both `MFEM_USE_SINGLE=YES` and `MFEM_USE_DOUBLE=NO` (see https://gitlab.com/petsc/petsc/-/merge_requests/7414) . Is this expected?

**Comment by tzanio** [comment]
> @tzanio I have added support for single precision. 

Thank you!

> When building using a `user.mk` file (the method used when PETSc builds MFEM) I have noticed that I have to specify both `MFEM_USE_SINGLE=YES` and `MFEM_USE_DOUBLE=NO` (see https://gitlab.com/petsc/petsc/-/merge_requests/7414) . Is this expected?

Can you just say `MFEM_PRECISION = single`? This should be translated to above

**Comment by v-dobrev** [comment]
If PETSc works with both single and double precision, we can add it to the list of supported external libraries here: #4207. Should I add it there? If yes, I'd like to add a bullet about how to build PETSc with single precision -- what is the option for that?

**Comment by stefanozampini** [comment]
> If PETSc works with both single and double precision, we can add it to the list of supported external libraries here: #4207. Should I add it there? If yes, I'd like to add a bullet about how to build PETSc with single precision -- what is the option for that?

Done

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Do we need to update `CHANGELOG`?

**Comment by tzanio** [comment]
Hi @stefanozampini,

We got the following failure in our internal testing with this PR:

```
    Parallel PETSc example [ srun -n 4 ex3p ... ]: FAILED  (1.67s 11124kB)
Options used:
   --mesh ../../data/amr-hex.mesh
   --refine-serial 1
   --refine-parallel 0
   --order 2
   --frequency 0.1
   --no-static-condensation
   --no-visualization
   --usepetsc
   --petscopts rc_ex3p_bddc
   --nonoverlapping
Number of finite element unknowns: 23496
Size of linear system: 23496
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Object is in wrong state
[0]PETSC ERROR: garray not present
[0]PETSC ERROR: See https://petsc.org/release/faq/ for trouble shooting.
[0]PETSC ERROR: Petsc Release Version 3.16.1, Nov 01, 2021 
[0]PETSC ERROR: Configure options --prefix=mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/petsc-3.16.1-tnxjua4ytj3uk7gdftktlw72jsqxgfvb --with-ssl=0 --download-c2html=0 --download-sowing=0 --download-hwloc=0 CFLAGS= FFLAGS= CXXFLAGS= --with-cc=/usr/tce/packages/mvapich2/mvapich2-2.3.6-gcc-10.3.1-magic/bin/mpicc --with-cxx=/usr/tce/packages/mvapich2/mvapich2-2.3.6-gcc-10.3.1-magic/bin/mpicxx --with-fc=/usr/tce/packages/mvapich2/mvapich2-2.3.6-gcc-10.3.1-magic/bin/mpif90 --with-precision=double --with-scalar-type=real --with-shared-libraries=1 --with-debugging=0 --with-openmp=0 --with-64-bit-indices=0 COPTFLAGS= FOPTFLAGS= CXXOPTFLAGS= --with-blaslapack-lib=/usr/lib64/libopenblas.so --with-x=0 --with-clanguage=C --with-cuda=0 --with-hip=0 --with-metis=1 --with-metis-include=mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/metis-5.1.0-fpd755h347uvwgfpme74sk43nkiji52v/include --with-metis-lib=mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/metis-5.1.0-fpd755h347uvwgfpme74sk43nkiji52v/lib/libmetis.so --with-hypre=1 --with-hypre-include=mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/hypre-2.23.0-v46fyqags55wiq47fl2f5kn6qlt2m36g/include --with-hypre-lib=mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/hypre-2.23.0-v46fyqags55wiq47fl2f5kn6qlt2m36g/lib/libHYPRE.so --with-parmetis=1 --with-parmetis-include=mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/parmetis-4.0.3-6vjtqw4hf3xa55ija2qlul34jpfwexxy/include --with-parmetis-lib=mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/parmetis-4.0.3-6vjtqw4hf3xa55ija2qlul34jpfwexxy/lib/libparmetis.so --with-superlu_dist=0 --with-ptscotch=0 --with-suitesparse=0 --with-hdf5=1 --with-hdf5-include=mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/hdf5-1.10.7-zq6uw5w3bp2upw3v2hu5n4cz2uv4m2h3/include --with-hdf5-lib=mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/hdf5-1.10.7-zq6uw5w3bp2upw3v2hu5n4cz2uv4m2h3/lib/libhdf5.so --with-zlib=1 --with-zlib-include=/usr/include --with-zlib-lib=/usr/lib64/libz.so --with-mumps=1 --with-mumps-include=mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/mumps-5.4.0-pmankjkollq7aq52563ecusaurydxrl5/include --with-mumps-lib="mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/mumps-5.4.0-pmankjkollq7aq52563ecusaurydxrl5/lib/libmumps_common.so mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/mumps-5.4.0-pmankjkollq7aq52563ecusaurydxrl5/lib/libsmumps.so mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/mumps-5.4.0-pmankjkollq7aq52563ecusaurydxrl5/lib/libcmumps.so mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/mumps-5.4.0-pmankjkollq7aq52563ecusaurydxrl5/lib/libdmumps.so mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/mumps-5.4.0-pmankjkollq7aq52563ecusaurydxrl5/lib/libzmumps.so mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/mumps-5.4.0-pmankjkollq7aq52563ecusaurydxrl5/lib/libpord.so" --with-trilinos=0 --with-fftw=0 --with-valgrind=0 --with-gmp=0 --with-libpng=0 --with-giflib=0 --with-mpfr=0 --with-netcdf=0 --with-pnetcdf=0 --with-moab=0 --with-random123=0 --with-exodusii=0 --with-cgns=0 --with-memkind=0 --with-p4est=0 --with-saws=0 --with-yaml=0 --with-hwloc=0 --with-libjpeg=0 --with-scalapack=1 --with-scalapack-lib=mfem/mfem-dependencies/mfem/uberenv_libs/linux-rhel8-broadwell/gcc-10.3.1/netlib-scalapack-2.1.0-3pfmcltei3legsk2rfzty5djwy57a6o5/lib/libscalapack.so --with-strumpack=0 --with-mmg=0 --with-parmmg=0 --with-tetgen=0
[0]PETSC ERROR: #1 MatConvert_XAIJ_IS() at mfem/mfem-dependencies/mfem/uberenv_libs/builds/dobrev1/spack-stage-petsc-3.16.1-tnxjua4ytj3uk7gdftktlw72jsqxgfvb/spack-src/src/mat/impls/is/matis.c:714
[0]PETSC ERROR: #2 MatConvert() at mfem/mfem-dependencies/mfem/uberenv_libs/builds/dobrev1/spack-stage-petsc-3.16.1-tnxjua4ytj3uk7gdftktlw72jsqxgfvb/spack-src/src/mat/interface/matrix.c:4435
[0]PETSC ERROR: #3 PCBDDCNedelecSupport() at mfem/mfem-dependencies/mfem/uberenv_libs/builds/dobrev1/spack-stage-petsc-3.16.1-tnxjua4ytj3uk7gdftktlw72jsqxgfvb/spack-src/src/ksp/pc/impls/bddc/bddcprivate.c:373
[0]PETSC ERROR: #4 PCSetUp_BDDC() at mfem/mfem-dependencies/mfem/uberenv_libs/builds/dobrev1/spack-stage-petsc-3.16.1-tnxjua4ytj3uk7gdftktlw72jsqxgfvb/spack-src/src/ksp/pc/impls/bddc/bddc.c:1655
[0]PETSC ERROR: #5 PCSetUp() at mfem/mfem-dependencies/mfem/uberenv_libs/builds/dobrev1/spack-stage-petsc-3.16.1-tnxjua4ytj3uk7gdftktlw72jsqxgfvb/spack-src/src/ksp/pc/interface/precon.c:1016
[0]PETSC ERROR: #6 KSPSetUp() at mfem/mfem-dependencies/mfem/uberenv_libs/builds/dobrev1/spack-stage-petsc-3.16.1-tnxjua4ytj3uk7gdftktlw72jsqxgfvb/spack-src/src/ksp/ksp/interface/itfunc.c:408
[0]PETSC ERROR: #7 KSPSolve_Private() at mfem/mfem-dependencies/mfem/uberenv_libs/builds/dobrev1/spack-stage-petsc-3.16.1-tnxjua4ytj3uk7gdftktlw72jsqxgfvb/spack-src/src/ksp/ksp/interface/itfunc.c:852
[0]PETSC ERROR: #8 KSPSolve() at mfem/mfem-dependencies/mfem/uberenv_libs/builds/dobrev1/spack-stage-petsc-3.16.1-tnxjua4ytj3uk7gdftktlw72jsqxgfvb/spack-src/src/ksp/ksp/interface/itfunc.c:1086
[0]PETSC ERROR: #9 void mfem::PetscLinearSolver::MultKernel(const mfem::Vector&, mfem::Vector&, bool) const() at linalg/petsc.cpp:3184
MFEM abort: Error in PETSc. See stacktrace above.
 ... in function: void mfem::PetscLinearSolver::MultKernel(const mfem::Vector&, mfem::Vector&, bool) const
 ... in file: linalg/petsc.cpp:3184
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
```

Can you please take a look?

**Comment by stefanozampini** [comment]
This was not tested before. Fixed here https://github.com/mfem/mfem/issues/4050. Can you use PETSc 3.21 for your internal testing? 3.16 is 3 years old

**Comment by adam-sim-dev** [comment]
> This was not tested before. See also https://github.com/mfem/mfem/issues/4050. Can you use PETSc 3.21 for your internal testing? 3.16 is 3 years old

Hi @stefanozampini , can you have a look at building mfem with CUDA and Intel oneAPI? I see petsc-3.21 adds "-fortlib" to the compiler options which causes the build failure.

**Comment by stefanozampini** [comment]
send an email to petsc-users@mcs.anl.gov

**Comment by tzanio** [comment]
> Can you use PETSc 3.21 for your internal testing? 3.16 is 3 years old

@v-dobrev, what do you think?

**Comment by v-dobrev** [comment]
> > Can you use PETSc 3.21 for your internal testing? 3.16 is 3 years old
> 
> @v-dobrev and @jandrej , what do you think?

It sounds like PETSc versions older than 3.21 will not work with this particular run, so we probably need to bump the required PETSc version to 3.21 in the INSTALL file. We can clarify the that older versions should also work (with the versions we require now) if one does not use the solver that was fixed in 3.21.

Updating the PETSc versions used in testing will require some work but that is a secondary concern -- the main question is if we want to require PETSc 3.21 and seems like the answer is yes.

Alternatively, in `petsc/ex3p`, if the version is older than 3.21 and the options ask for the solver that was fixed in 3.21, we can print explanatory message and return the `MFEM_SKIP_RETURN_VALUE`. This way `make test` will print "SKIP" status and not fail.

@stefanozampini, @tzanio, @jandrej, @sebastiangrimberg, and others, what is you preference? Or do you have another suggestion?

**Comment by sebastiangrimberg** [comment]
I agree @v-dobrev  and suggest just updating the minimum required built to 3.21. I don't think that puts any substantial burden onto users. Not sure how much work it entails in the CI system, but if the underlying dependency builds just use Spack (which it looks like they do from the `.gitlab` directory), then it doesn't look like too much work to make the update. Let me know if I can help in any way here.

**Comment by stefanozampini** [comment]
@v-dobrev I'm now skipping the 3D test if the PETSc version is not supported.

**Comment by tzanio** [comment]
@stefanozampini, can you change the required minimum version for PETSc to 3.21 in `INSTALL` and mention it in `CHANGELOG`?

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
> Can you change the required minimum version for PETSc to 3.21 in INSTALL and mention it in CHANGELOG?

ping: @stefanozampini

**Comment by stefanozampini** [comment]
> > Can you change the required minimum version for PETSc to 3.21 in INSTALL and mention it in CHANGELOG?
> 
> ping: @stefanozampini

@tzanio go ahead and do it

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Pull request merged**